### DISCUSSION TITLE
Coverage map of non-flat surface

### DISCUSSION THREAD

**Discussion by Jklein64** [NVlabs/sionna#610] [Q&A] [answered]
Is it possible to generate a coverage map for a 2D surface that isn’t flat, such as a mesh terrain specified by a heightmap? 

I’ve experimented with the Sionna coverage map function and have only been able to change the orientation of the plane on which the coverage map is computed, as opposed to computing the coverage at different heights. Below is an example of why this behavior is problematic for us:

![image](https://github.com/user-attachments/assets/19b97af3-5f57-41da-b0ea-319212eb5ced)

Our current workaround idea is to place many (hundreds or thousands) of receivers scattered across the terrain mesh, and then use the path losses from the transmitter to each receiver to get the data we need. This will be slow, and it would be helpful to be able to evaluate the coverage map function at the varying heights instead. Is there something fundamental about the coverage map computation algorithm that makes this difficult?

**Comment by faycalaa** [comment]
Hi @Jklein64,

It is possible to set the position of the center of the coverage map using the `cm_center` parameter of [Scene.coverage_map](https://nvlabs.github.io/sionna/api/rt.html#sionna.rt.Scene.coverage_map). However, it is currently not possible to compute coverage maps for non-planar surfaces. The existing implementation assumes that the measurement surface is flat when calculating the contribution to the coverage map of a path intersecting the measurement plane.

In principle, it would be possible to consider non-planar measurement surfaces, but we currently do not have plans to implement this feature.

You could consider computing multiple smaller coverage maps, each with different positions and orientations, to approximate the shape of the terrain.

**Reply by Mgomez-01** [reply]
@faycalaa  Using an offset from the terrain, like 1.5m above it, is something that I would very much like to have for the work I am doing at the University of Utah with P.O.W.D.E.R. We have many sites on our campus and the campus itself is on a mountainside with ~200m of height difference between one side and the other, so using a flat map really offers us no benefit. 

I have looked around the source code a bit and see that it uses a Mitsuba Shape for the rectangle used for the coverage_map. I understand that this is not something that you have plans to implement, but is there someone who could give me a bit of guidance as for the best way to try and work this into the solver? I suspect that if we replace the shape with one made from the Terrain.ply file, we could get something that would sort of work, but I am still working on my familiarity with the source code. Any help or guidance you could provide would be great.

**Reply by faycalaa** [reply]
Hi @Mgomez-01,

You can in principle load a mesh using Mitsuba instead of using a rectangle, and then use it as measurement surface to compute the coverage map.

The orientation of the intersected primitive would need to be considered when computing the contribution to the coverage map of a ray intersecting the measurement surface.

This could be a feature that we could implement in a future release.

**Reply by Jklein64** [reply]
Hi @Mgomez-01! My team is currently working on modifications to Sionna that would allow for computation of coverage maps on non-planar surfaces represented by heightmaps, which seems like it would fit your use case as well. Our current approach aligns well with what @faycalaa mentioned, and we’ve already worked out the math to account for different orientations and cell areas. We’ll post an update here once we get something working.

**Reply by Mgomez-01** [reply]
@faycalaa Thanks for your reply, 
That workflow is what I was thinking, and I will keep looking into the source to see where it could fit best. Also excited to see that @Jklein64 is already making progress on this problem.

**Reply by Mgomez-01** [reply]
@Jklein64 That definitely sounds like it fits my use case! I am excited to see it once you have something working.

**Reply by redcoatwright** [reply]
@Jklein64 have you managed to make any progress on the issue of using a non-planar coverage map geometry? If not, I'll be diving into the source code as well to try to get it working.

**Reply by Jklein64** [reply]
@redcoatwright our initial approach had some pretty hefty drawbacks (such as requiring edits to most of the coverage map solver), but we recently transitioned to another approach based on textures that seems much more promising. Here are some images from recent experiments: 

| ![image](https://github.com/user-attachments/assets/de140b16-b6c9-4050-b245-68fa2511b25c) | ![image](https://github.com/user-attachments/assets/9951a679-6d93-4a76-8cb1-69b898468d03) | ![image](https://github.com/user-attachments/assets/d38a1eca-d641-4da6-9341-8a585fdf2239) |
| --- | --- | --- |

I plan to integrate this into Sionna and the coverage map solver soon. This method should require much less codebase updates than the previous approach, so fingers crossed it all falls into place nicely 🤞

**Reply by redcoatwright** [reply]
@Jklein64 is there any chance I can check out your fork where you're working on this? Trying to see if I can make this work for my use-case

**Reply by Jklein64** [reply]
@redcoatwright I realized I need the surface area of each cell as part of the coverage map computation, which is hard to compute for non-flat measurement surfaces. I've designed my own algorithm for this and am currently testing it, but it'll be a bit before everything is integrated together. I'll let you know once I have a fork you can look at!

**Reply by jhoydis** [reply]
Hi all,

Please note that the implementation of coverage maps has significantly changed in the new Sionna 1.0. Non-flat coverage maps are still not supported, but it is definitely an interesting feature we should consider for a future release.

**Comment by Jklein64** [comment, marked as answer]
Support for arbitrary mesh measurement surfaces in Radio Maps has been added in [Sionna 1.1.0](https://github.com/NVlabs/sionna/releases/tag/v1.1.0). Its implementation uses each mesh face as a “cell” instead of generalizing the grid of cells method (from planar Radio Maps) to work with non-flat measurement surfaces. This works, but the results are not exposed as a grid, and it's difficult to adjust the resolution of cells (since there is no analog to the `cell_size` parameter).

These two features, (1) a 2D path gain tensor and (2) a `cell_size` parameter, were very important to us. To handle this, I created another Radio Map solver that takes Digital Elevation Maps (DEMs, aka heightmaps) as input and evaluates the path loss on a "grid" draped over the heightmap. See [this PR](https://github.com/NVlabs/sionna-rt/pull/25) for more info.

Hope this helps! @redcoatwright @Mgomez-01
